File: executable-operator/executable-basic/src/main/java/fdu/daslab/executable/basic/utils/TopoTraversal.java

package fdu.daslab.executable.basic.utils;

import fdu.daslab.executable.basic.model.OperatorBase;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author 陈齐翔
 * @version 1.0
 * @since 2020/9/28 11:01 下午
 */
public class TopoTraversal {
    Queue<OperatorBase> queue = new LinkedList<>();

    public TopoTraversal(List<OperatorBase> heads) {
        queue.addAll(heads);
    }

    public OperatorBase nextOpt() {
        return queue.poll();
    }

    public void updateInDegree(OperatorBase opt, int delta) {
        opt.updateInDegree(delta);
        if (opt.getInDegree() <= 0) {
            queue.add(opt);
        }
    }

    public boolean hasNextOpt() {
        return !this.queue.isEmpty();
    }
}

File: executable-operator/executable-pytorch/main/utils/TopoTraversal.py
from queue import SimpleQueue
"""
@ProjectName: CLIC
@Time       : 2020/11/25 下午1:05
@Author     : zjchen
@Description: 
"""


class TopoTraversal:
    def __init__(self, headList):
        self.queue = SimpleQueue()
        for head in headList:
            self.queue.put(head)

    def nextOpt(self):
        return self.queue.get()

    def updateInDegree(self, opt, delta):
        opt.updateInDegree(delta)
        if opt.getInDegree() <= 0:
            self.queue.put(opt)

    def hasNextOpt(self):
        return not self.queue.empty()
